Set 0600 permissions on `credentials.toml`

wfchandler opened the pull request:
Currently we create the `credentials.toml` file with 0644 permission on Unix. Given that this file contains sensitive data, 0600 is a more appropriate permission.

On Windows my rather vague understanding is that the `%USERNAME%` directory we default to is only accessible by that user and admins. There are additional settings like ACLs we could tweak, but these are not accessible from the Rust standard library and probably more trouble than they're worth.

We deliberately do not set `config.toml` as 0600 as this does not contain any sensitive information.

wfchandler committed:
Set 0600 permissions on credentials.toml

Currently we create the `credentials.toml` file with 0644 permission on
Unix. Given that this file contains sensitive data, 0600 is a more
appropriate permission.

On Windows my rather vague understanding is that the `%USERNAME%`
directory we default to is only accessible by that user (and admins).
There are additional settings like ACLs we could tweak, but these are
not accessible from the Rust standard library and probably more trouble
than they're worth.

We deliberately do not set `config.toml` as 0600 as this does not
contain any sensitive information.

wfchandler commented:
After clearing my config and re-authing:

```sh
$ ls -l ~/.config/oxide
-rw-r--r--@  1 wfc    26B Oct  3 01:29 config.toml
-rw-------@  1 wfc   173B Oct  3 01:29 credentials.toml
```

ahl commented on cli/src/cmd_auth.rs:
```diff
@@ -535,6 +544,23 @@ impl CmdAuthStatus {
 
+/// Create a file. On Unix set permissions to 0600.
+fn create_private_file(path: &Path) -> Result<File> {
+    let file = File::create(path)?;
```
I think this leaves a race where one could open the file before the permissions were changed.

wfchandler replied:
Good callout, I've switched to `OpenOptions` to set the mode on creation.

ahl reviewed (changes requested):
I think we want to use https://doc.rust-lang.org/std/os/unix/fs/trait.OpenOptionsExt.html#tymethod.mode

wfchandler committed:
Set mode when creating file

Using `File::create` will create a new file with 0644 permissions, which
potentially leaves an interval where another user could access the file
before we updated it to 0600.

Switch to using `OpenOptions` so we can directly create the file with
the correct permissions. However, in cases where the file already
exists, `OpenOptions::mode` will not overwrite the mode of an existing
file, so we retain the subsequent `fstat` and `fchmod` calls to handle
that scenario.

ahl commented on cli/src/cmd_auth.rs:
```diff
@@ -535,6 +544,36 @@ impl CmdAuthStatus {
+        .truncate(true)
+        .open(path)?;
+
+    // `OpenOptions::mode` will not modify the permissions of an existing file.
```
If someone changes the permissions do we really want to override them? Or do you view this as temporary to fix up existing users?

wfchandler replied:
The latter, but perhaps that's a bad idea. Do you have a strong preference?

wfchandler replied:
[ssh](https://www.man7.org/linux/man-pages/man1/ssh.1.html#FILES) will ignore private keys with overly broad permissions. We could refuse to read a non-600 file, but that would effectively log out all users, so perhaps printing a warning and proceeding is a better option.

ahl replied:
Agreed: warn and proceed

wfchandler replied:
Added the following warning, which will trigger on any call that constructs a `Context`.

```
WARNING: 644 permissions on "<CREDS_FILE>" may allow other users to access your login credentials.
```

ahl commented on cli/tests/test_auth.rs:
```diff
@@ -123,11 +146,13 @@ fn test_auth_login_first() {
         "tests/data/test_auth_login_first_config.toml",
         &read_to_string(config_dir.join("config.toml")).unwrap(),
     );
+    assert_mode(&config_dir.join("config.toml"), 0o100644);
```
This will depend on the umask I think

wfchandler replied:
Good point, I've updated the tests to take this into account.

ahl commented on cli/tests/test_auth.rs:
```diff
@@ -155,7 +180,9 @@ fn test_auth_login_existing_default() {
 
     let temp_dir = tempfile::tempdir().unwrap().into_path();
     write_first_creds(&temp_dir);
+    assert_mode(&temp_dir.join("credentials.toml"), 0o100644);
```
600?

wfchandler replied:
644 is correct here since we're using the `write_first_creds` helper to initially create the file. I think keeping that as-is is helpful, as it lets us confirm that we update the permissions on existing files.

ahl replied:
I think it's worth a comment in that case

wfchandler replied:
Since we're no longer updating the permissions of existing files, I updated the helper to create the file as 600.

wfchandler committed:
Take umask into consideration in tests

We may receive spurious test failures if the system running the test
uses a different `umask` than out own. Apply the host's `umask` against
the expected mode to ensure we're performing a valid comparison.

wfchandler committed:
Warn if credentials are accessible to other users

Rather than forcing the permissions of `credentials.toml` to 0600, print
a warning along the lines of the message that OpenSSH displays when a
private key's permissions are not as expected.

While we're at it, fix a typo in sdk/src/auth.rs.
